### Let's write an elementary tokenizer that uses words as tokens.

We will use Mark Twain's _Life On The Mississippi_ as a test bed. The text is in the accompanying file 'Life_On_The_Mississippi.txt'

Here's a not-terribly-good such tokenizer:

In [2]:
wdict = {}
with open('Life_On_The_Mississippi.txt', 'r') as L:
    line = L.readline()
    nlines = 1
    while line:

        words = line.split()
        for word in words:
            if wdict.get(word) is not None:
                wdict[word] += 1
            else:
                wdict[word] = 1
        line = L.readline()
        nlines += 1

nitem = 0 ; maxitems = 100
for item in wdict.items():
    nitem += 1
    print(item)
    if nitem == maxitems: break


('\ufeffThe', 1)
('Project', 79)
('Gutenberg', 22)
('eBook', 4)
('of', 4469)
('Life', 5)
('on', 856)
('the', 8443)
('Mississippi', 104)
('This', 127)
('ebook', 2)
('is', 1076)
('for', 1017)
('use', 34)
('anyone', 4)
('anywhere', 8)
('in', 2381)
('United', 36)
('States', 26)
('and', 5692)
('most', 119)
('other', 223)
('parts', 5)
('world', 40)
('at', 676)
('no', 325)
('cost', 18)
('with', 1053)
('almost', 37)
('restrictions', 2)
('whatsoever.', 2)
('You', 92)
('may', 85)
('copy', 12)
('it,', 199)
('give', 67)
('it', 1382)
('away', 107)
('or', 561)
('re-use', 2)
('under', 112)
('terms', 22)
('License', 8)
('included', 2)
('this', 591)
('online', 4)
('www.gutenberg.org.', 4)
('If', 85)
('you', 813)
('are', 361)
('not', 680)
('located', 9)
('States,', 8)
('will', 287)
('have', 557)
('to', 3518)
('check', 4)
('laws', 13)
('country', 50)
('where', 152)
('before', 150)
('using', 10)
('eBook.', 2)
('Title:', 1)
('Author:', 1)
('Mark', 2)
('Twain', 2)
('Release', 1)
('date:', 1)
('July', 7)
('1

This is unsatisfactory for a few reasons:

* There are non-ASCII (Unicode) characters that should be stripped (the so-called "Byte-Order Mark" or BOM \ufeff at the beginning of the text);

* There are punctuation marks, which we don't want to concern ourselves with;

* The same word can appear capitalized, or lower-case, or with its initial letter upper-cased, whereas we want them all to be normalized to lower-case.

Part 1 of this assignment: insert code in this loop to operate on the str variable 'line' so as to fix these problems before 'line' is split into words.

A hint to one possible way to do this: use the 'punctuation' character definition in the Python 'string' module, the 'maketrans' and 'translate' methods of Python's str class, to eliminate punctuation, and the regular expression ('re') Python module to eliminate any Unicode---it is useful to know that the regular expression r'[^\x00-x7f]' means "any character not in the vanilla ASCII set.

Part 2: Add code to sort the contents of wdict by word occurrence frequency.  What are the top 100 most frequent word tokens?  Adding up occurrence frequencies starting from the most frequent words, how many distinct words make up the top 90% of word occurrences in this "corpus"?

For this part, the docs of Python's 'sorted' and of the helper 'itemgetter' from 'operator' reward study.

Write your modified code in the cell below.

In [9]:
import re
from string import punctuation

mytable = str.maketrans(punctuation, ' '*len(punctuation))
wdict = {}
with open('Life_On_The_Mississippi.txt', 'r') as L:
    line = L.readline()
    nlines = 1
    while line:
        line = line.translate(mytable)
        #line.strip(r'[^\x00-x7f]')
        words = line.split()
        for word in words:
            word = re.sub(r'[^\x00-x7f]','', word)  
            word = word.lower()
            if wdict.get(word) is not None:
                wdict[word] += 1
            else:
                wdict[word] = 1
        line = L.readline()
        nlines += 1

        
wdict = {k: v for k, v in sorted(wdict.items(), key=lambda item: item[1], reverse = True)}
total_sum = sum([el[1] for el in wdict.items()])

occur = 0
nitem = 0 ; maxitems = 100
for item in wdict.items():
    nitem += 1
    occur += item[1]
    print(item, occur/total_sum*100)

    if nitem == maxitems: break


('the', 10147) 6.657655943468647
('and', 6032) 10.615375530637552
('of', 4546) 13.59809987468096
('a', 4237) 16.378082946768934
('to', 3633) 18.761769163643045
('in', 2622) 20.482117432468787
('it', 2404) 22.059431405869656
('i', 2376) 23.618374001876504
('was', 2126) 25.013286442579602
('that', 1789) 26.18708623393325
('he', 1448) 27.137148893452572
('is', 1163) 27.900217175925622
('for', 1123) 28.637040633550072
('with', 1096) 29.35614883440172
('ou', 990) 30.005708249404574
('but', 988) 30.653955423164998
('his', 967) 31.288424063879905
('on', 962) 31.919612101488738
('had', 962) 32.550800139097575
('as', 889) 33.13409137135771
('this', 795) 33.65570726522364
('b', 761) 34.15501505796826
('at', 755) 34.65038612698558
('all', 738) 35.134603145442256
('not', 735) 35.616851802035285
('one', 730) 36.09581985552224
('s', 706) 36.55904101410003
('an', 696) 37.01570096646567
('there', 675) 37.4585823857858
('be', 635) 37.87521898025733
('were', 627) 38.28660660975914
('m', 625) 38.69668199

In [10]:
occur = 0
nitem = 0 ; maxitems = 100
for item in wdict.items():
    nitem += 1
    occur += item[1]
    #print(item, occur/total_sum*100)

    if occur/total_sum*100 > 90: print(nitem); break

3027


3027 words make the 90% of occurance in the corpus